wire/wire.go

```diff
@@ -13,22 +13,60 @@
 // limitations under the License.
 
 // Package wire contains directives for Wire code generation.
+// For an overview of working with Wire, see the user guide at
+// https://github.com/google/go-cloud/blob/master/wire/README.md
+//
+// The directives in this package are used as input to the Wire code
+// generation tool. The entry point of Wire's analysis are injector functions:
+// function templates denoted by only containing a call to Build. Build names
+// a set of providers and the Wire code generation tool builds a directed
+// acylic graph of the providers' output types. The generated generated code
+// will fill in the function template by calling the providers from the
+// provider set.
 package wire
 
 // ProviderSet is a marker type that collects a group of providers.
 type ProviderSet struct{}
 
-// NewSet creates a new provider set that includes the providers in
-// its arguments. Each argument is either an exported function value,
-// an exported struct (zero) value, a provider set, a call to Bind,
-// a call to Value, or a call to InterfaceValue.
+// NewSet creates a new provider set that includes the providers in its
+// arguments. Each argument is a function value, a struct (zero) value, a
+// provider set, a call to Bind, a call to Value, or a call to InterfaceValue.
+//
+// Passing a function value to NewSet declares that the function's first
+// return value type will be provided by calling the function. The arguments
+// to the function will come from the providers for their types. As such, all
+// the parameters must be of non-identical types. The function may optionally
+// return an error as its last return value and a cleanup function as the
+// second return value. A cleanup function must be of type func() and is
+// guaranteed to be called before the cleanup function of any of the
+// provider's inputs. If any provider returns an error, the injector function
+// will call all the appropriate cleanup functions and return the error from
+// the injector function.
+//
+// Passing a struct value of type S to NewSet declares that both S and *S will
+// be provided by creating a new value of the appropriate type by filling in
+// each field of S using the provider of the field's type.
```

Comment:
each exported field?

Reply:
Doesn't have to be, if S is in the same package as the injector.